src/main/java/jenkins/scm/api/SCMFileSystem.java

```diff
@@ -527,4 +556,27 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head,
             return build(owner, source.build(head, rev), rev);
         }
     }
+
+    public abstract static class Builder2 extends Builder {
+
+        /**
+         * Given a {@link SCM} this should try to build a corresponding {@link SCMFileSystem} instance that
+         * reflects the content at the specified {@link SCMRevision}. If the {@link SCM} is supported but not
+         * for a fixed revision, best effort is acceptable as the most capable {@link SCMFileSystem} will be returned
+         * to the caller.
+         *
+         * @param owner the owner of the {@link SCM}
+         * @param scm the {@link SCM}.
+         * @param rev the specified {@link SCMRevision}.
+         * @return the corresponding {@link SCMFileSystem} or {@code null} if this builder cannot create a {@link
+         * SCMFileSystem} for the specified {@link SCM}.
+         * @throws IOException          if the attempt to create a {@link SCMFileSystem} failed due to an IO error
+         *                              (such as the remote system being unavailable)
+         * @throws InterruptedException if the attempt to create a {@link SCMFileSystem} was interrupted.
+         */
+        @CheckForNull
```

Comment:
in GitSCM I can make GitSCMFileSystem to override from this builder instead of the old and pass the Run<?,?>